### PR TITLE
Add support for response writing to RequestDelegateGenerator

### LINKED ISSUE #45524: Generate route handler-endpoints at compile time via source generator
Currently, RequestDelegates for associated route handler-based endpoints are generated at runtime using LINQ expressions. This is a process that requires a fair amount of reflection. We can leverage source generation to generate these endpoints via static analysis.

Some of this work has been prototyped in [uController](https://github.com/davidfowl/uController) and will need to be formalized as part of the official project.

**Generator Basics**
- [x] Add RequestDelegateGenerator scaffold to aspnetcore repo
- [x] Respect `EnableRequestDelegateGenerator` flag in Web SDK
- [x] Refactor code to support marking `GeneratedRouteBuilderExtensions` as `file class`
- [ ] Set up code and tests for emitting diagnostics from the generator
- [x] Support for a RouteEndpointDataSource implementation that supports a func for populating metadata
- [x] Support for an RouteEndpointDataSource implementation that allows customizing what RDF is called to produce filtered delegate calls

**Invoking Handlers**

- [ ] Support generating correct delegate type signature for a given handler
- [ ] Support emitting correct  generic-`EndpointFilterInvocationContext` for a given filtered invocation

### LINKED ISSUE #46276: Support response writing in RequestDelegateGenerator
**Responses**
- [ ] Support emitting correct `Response.WriteX` call given return type
  - [ ] Async delegates
  - [ ] Delegates returning `Task`/`ValueTask`
  - [ ]  Support serializing polymorphic responses with STJ

### PR DESCRIPTION
**Pull request by captainsafia**
Addresses https://github.com/dotnet/aspnetcore/issues/46276 and resolves #45524.

- Share WellKnownTypes infrastructure between analyzers and generator
- Emit diagnostics from RequestDelegateGenerator for unsupported syntax
- Add support for writing more response types to generator
- Dedupe strongly-typed endpoints generated from the same delegate type
- Use RequestDelegateGenerator in MinimalSample
- Fix nullability warnings in generated source when referencing `RequestDelegateFactoryOptions.EndpointBuilder`

### CODE REVIEW

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs:8]: Does this need to be nested?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt
Comment (JamesNK) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt:78]: super nit: Is there a way to not start the type with a blank line? My OCD is triggered 😄

File: src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.Generators.csproj
Comment (JamesNK) [inline src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.Generators.csproj:28]: I think sharing code is ok for now, but doing this results in two caches. I think we should investigate whether we can have a shared library project output to the analyzers directory that generators, analyzers, and code fixers can reference.

For route pattern parsing, there is a lot more code, and we don't want to duplicate parsing and caching route patterns twice (once in analyzer assembly, once in generators assembly)

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt:78]: Hm, I think we might be able to remove the newline in the interpolated string `{GetGenericThunks()}{GetThunks()}` and change the strings emitted from those methods to get around it.

In the long-term, we might be able to take advantage of the `SyntaxTree.NormalizeWhitespace` feature to correct the indentation for the outputted source. I wasn't able to use it successfully here because it has a really hard time formatting the Dictionary of source key to tuples of funcs pairing. As we evolve the code, we might be able to get rid of this (e.g. once the interceptors feature lands) and we can do all the formatting at once instead of doing the bookkeeping for the indents here.

A related issue on the Roslyn side is: https://github.com/dotnet/roslyn/issues/43821

File: src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.Generators.csproj
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.Generators.csproj:28]: > I think sharing code is ok for now, but doing this results in two caches. I think we should investigate whether we can have a shared library project output to the analyzers directory that generators, analyzers, and code fixers can reference.

This is an interesting proposal. @chsienki do you happen to know if there is a pattern for sharing a symbol cache across analyzers/generators in the same compilation process?

> For route pattern parsing, there is a lot more code, and we don't want to duplicate parsing and caching route patterns twice (once in analyzer assembly, once in generators assembly)

Roger that! I captured that in https://github.com/dotnet/aspnetcore/issues/46342 so we can start discussing what that might look like.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs:8]: Not necessarily. It was largely a stylistic choice to put the associated data in the same class. To minimize the code delta, I ended up using an alias in the existing analyzers/codefixers but it felt nicer to capture these types in class instead of having them on the top-level namespace.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
Comment (eerhardt) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:89]: `EmitResponseWritingCall` handles `IsAwaitable()`. So I think this can be refactored to just:

```suggestion
            code.WriteLine(endpoint.Response.IsAwaitable
                ? "var result = await handler();"
                : "var result = handler();");
            code.WriteLine(endpoint.EmitResponseWritingCall());
```

File: src/Http/Http.Extensions/gen/GeneratorSteps.cs
Comment (eerhardt) [inline src/Http/Http.Extensions/gen/GeneratorSteps.cs:7]: Is it weird that the name of the variable and the string doesn't match?

File: src/Http/Http.Extensions/gen/GeneratorSteps.cs
Comment (eerhardt) [inline src/Http/Http.Extensions/gen/GeneratorSteps.cs:8]:
```suggestion
    internal const string EndpointsModel = nameof(EndpointModel);
    internal const string EndpointsWithoutDiagnosicsStep = nameof(EndpointsWithoutDiagnostics);
```

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs
Comment (eerhardt) [inline src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:133]: We could use the [Microsoft.Bcl.HashCode](https://nuget.org/packages/Microsoft.Bcl.HashCode) package here.

```suggestion
                        return HashCode.Combine(
                            SymbolEqualityComparer.Default.GetHashCode(endpoint.Response.ResponseType),
                            endpoint.Response.IsAwaitable,
                            endpoint.Response.IsVoid,
                            endpoint.HttpMethod);
```

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs
Comment (halter73) [inline src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:119]:
```suggestion
                    if (a.Response.WrappedResponseType == b.Response.WrappedResponseType && 
```

The call `EmitHandlerDelegateType()` uses the `WrappedResponseType`, so we need to compare those. Currently, the test case below only produces a MapGet overload for `Func<Task<string>>` and not `Func<ValueTask<string>>`.

```csharp
    [Fact]
    public async Task Multiple_MapAction_NoParam_TaskAndValueTaskReturn()
    {
        var source = """
app.MapGet("/en", () => Task.FromResult("Hello world!"));
app.MapGet("/es", () => new ValueTask<string>("Hola mundo!"));
""";
        var (_, compilation) = RunGenerator(source);

        await VerifyAgainstBaselineUsingFile(compilation);
    }
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryAddAnalyzer.cs
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryAddAnalyzer.cs:12]: Nit: Can we put this with the other usings?

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs:8]: Or can we just go back to not nesting this in another class so we can get rid of all the extra `using WellKnownType = WellKnownTypeData.WellKnownType;` lines?

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs
Comment (eerhardt) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs:80]: > but it's unavailable in netstandard2.0

https://www.nuget.org/packages/Microsoft.Bcl.HashCode makes it available.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs
Comment (halter73) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs:8]: Wouldn't leaving this unnested minimize the code delta by removing the `using WellKnownType = WellKnownTypeData.WellKnownType;`?

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs:80]: > https://www.nuget.org/packages/Microsoft.Bcl.HashCode makes it available.

Nice! It looks like we have this referenced in our dependencies already as well.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
Comment (mitchdenny) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:66]: When @captainsafia was getting @brunolins16 up to speed yesterday we were discussing the ```CodeWriter``` in general. My feedback was that using the CodeWriter rather than simple string concatenation/interpolation actually made it harder to reason over the generated code. I'd rather see us drop the CodeWriter (at the expense of indentation) and then either do some code formatting at the end before handing it back to Roslyn or just leave it a little messy. We can also do a Format Document on the generated code after build if we need to read it.

**Comment by captainsafia** [comment]
🆙 📅 
- Removed `CodeWriter` in favor of string literals with indentation in code sources
- Added incrementality tests to validate behavior if Endpoint's comparable implementation
- Added and consume `Microsoft.Bcl.HashCode` API as a dependency
- Tweaked smaller pieces of feedback

File: src/Http/samples/MinimalSample/MinimalSample.csproj
Comment (captainsafia) [reply src/Http/samples/MinimalSample/MinimalSample.csproj:7]: I don't use Visual Studio so I'm not able to take advantage of any shortcuts there.

> Do we need this all the time?

I found it helpful to have on while debugging as it makes it easier to examine the generated code in another format. There's no con to having it so I figured it was helpful.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs]: Ooof! Good catch on the `||` vs `&&`. I added incrementality tests that'll catch things like this in the future.

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs]: Good scenario! I added it to the tests.

**Review by halter73 — APPROVED** [review]
I've got a draft PR based on this branch open at #46407. Now I just want to minimize my merge conflicts! 😆 Looks good though!

**Comment by captainsafia** [comment]
Merging to unblock forthcoming PRs. Happy to address any other feedback in a followup PR. :shipit:

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
Comment (eerhardt) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs]: The Regex source generator uses `IndentedTextWriter`. Seems to work really well there. You just `writer.Indent++` and `--` when you are done.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L30

File: src/Http/samples/MinimalSample/MinimalSample.csproj
Comment (eerhardt) [reply src/Http/samples/MinimalSample/MinimalSample.csproj:7]: > I don't use Visual Studio

image

**Pull request merged**
